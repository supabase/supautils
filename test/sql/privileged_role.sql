-- non-superuser privileged role
set role privileged_role;
\echo

-- can set privileged_role_allowed_configs
set session_replication_role to 'replica';
begin;
  set local session_replication_role to 'origin';
commit;
reset session_replication_role;
\echo

-- non-superuser non-privileged role cannot set privileged_role_allowed_configs
set role rolecreator;

set session_replication_role to 'replica';
begin;
  set local session_replication_role to 'origin';
commit;
reset session_replication_role;

set role privileged_role;
\echo

-- superuser can set privileged_role_allowed_configs
set role postgres;

set session_replication_role to 'replica';
begin;
  set local session_replication_role to 'origin';
commit;
reset session_replication_role;

set role privileged_role;
\echo

-- can set privileged_role_allowed_configs for a role
create role r;
alter role r set session_replication_role to 'replica';
drop role r;
\echo

-- can manage bypassrls role attribute
create role r bypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
alter role r nobypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
alter role r bypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
drop role r;
\echo

-- non-superuser non-privileged role cannot manage bypassrls role attribute
set role rolecreator;

-- the error message changed in PG14
do $$
begin
  create role r bypassrls;
  exception when insufficient_privilege then null;
end;
$$ language plpgsql;

create role r;
alter role r nobypassrls;
alter role r bypassrls;
drop role r;

set role privileged_role;
\echo

-- superuser can manage bypassrls role attribute
set role postgres;

create role r bypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
alter role r nobypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
alter role r bypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
drop role r;

set role privileged_role;
\echo

-- regression: https://github.com/supabase/supautils/issues/34
create role tmp;
alter role tmp;
