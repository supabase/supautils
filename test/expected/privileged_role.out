-- non-superuser privileged role
set role privileged_role;
\echo

-- can set privileged_role_allowed_configs
set session_replication_role to 'replica';
begin;
  set local session_replication_role to 'origin';
commit;
reset session_replication_role;
\echo

-- non-superuser non-privileged role cannot set privileged_role_allowed_configs
set role rolecreator;
set session_replication_role to 'replica';
ERROR:  permission denied to set parameter "session_replication_role"
begin;
  set local session_replication_role to 'origin';
ERROR:  permission denied to set parameter "session_replication_role"
commit;
reset session_replication_role;
ERROR:  permission denied to set parameter "session_replication_role"
set role privileged_role;
\echo

-- superuser can set privileged_role_allowed_configs
set role postgres;
set session_replication_role to 'replica';
begin;
  set local session_replication_role to 'origin';
commit;
reset session_replication_role;
set role privileged_role;
\echo

-- can set privileged_role_allowed_configs for a role
create role r;
alter role r set session_replication_role to 'replica';
drop role r;
\echo

-- can manage bypassrls role attribute
create role r bypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
 rolbypassrls 
--------------
 t
(1 row)

alter role r nobypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
 rolbypassrls 
--------------
 f
(1 row)

alter role r bypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
 rolbypassrls 
--------------
 t
(1 row)

drop role r;
\echo

-- non-superuser non-privileged role cannot manage bypassrls role attribute
set role rolecreator;
-- the error message changed in PG14
do $$
begin
  create role r bypassrls;
  exception when insufficient_privilege then null;
end;
$$ language plpgsql;
create role r;
alter role r nobypassrls;
ERROR:  must be superuser to change bypassrls attribute
alter role r bypassrls;
ERROR:  must be superuser to change bypassrls attribute
drop role r;
set role privileged_role;
\echo

-- superuser can manage bypassrls role attribute
set role postgres;
create role r bypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
 rolbypassrls 
--------------
 t
(1 row)

alter role r nobypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
 rolbypassrls 
--------------
 f
(1 row)

alter role r bypassrls;
select rolbypassrls from pg_roles where rolname = 'r';
 rolbypassrls 
--------------
 t
(1 row)

drop role r;
set role privileged_role;
\echo

-- regression: https://github.com/supabase/supautils/issues/34
create role tmp;
alter role tmp;
